File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/SetCameraFocusPointActionTest.java
/*
 * Catroid: An on-device visual programming system for Android devices
 * Copyright (C) 2010-2023 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * An additional term exception under section 7 of the GNU Affero
 * General Public License, version 3, is available at
 * http://developer.catrobat.org/license_additional_term
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.catrobat.catroid.test.content.actions;

import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.scenes.scene2d.Action;
import com.badlogic.gdx.utils.GdxNativesLoader;

import org.catrobat.catroid.ProjectManager;
import org.catrobat.catroid.content.ActionFactory;
import org.catrobat.catroid.content.Project;
import org.catrobat.catroid.content.Sprite;
import org.catrobat.catroid.content.WhenScript;
import org.catrobat.catroid.content.actions.ScriptSequenceAction;
import org.catrobat.catroid.formulaeditor.Formula;
import org.catrobat.catroid.stage.CameraPositioner;
import org.catrobat.catroid.stage.StageActivity;
import org.catrobat.catroid.stage.StageListener;
import org.catrobat.catroid.test.MockUtil;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import org.catrobat.catroid.koin.CatroidKoinHelperKt;
import org.koin.core.module.Module;

import java.util.Collections;
import java.util.List;

import static junit.framework.Assert.assertEquals;

import static org.koin.java.KoinJavaComponent.inject;

@RunWith(PowerMockRunner.class)
@PrepareForTest({GdxNativesLoader.class})
public class SetCameraFocusPointActionTest {

	@Rule
	public final ExpectedException exception = ExpectedException.none();

	private Sprite sprite;

	final List<Module> dependencyModules =
			Collections.singletonList(CatroidKoinHelperKt.getProjectManagerModule());

	@Before
	public void setUp() {
		String projectName = "testProject";
		Project project = new Project(MockUtil.mockContextForProject(dependencyModules), projectName);
		sprite = new Sprite("sprite");
		sprite.addScript(new WhenScript());
		project.getDefaultScene().addSprite(sprite);
		inject(ProjectManager.class).getValue().setCurrentProject(project);

		PowerMockito.mockStatic(GdxNativesLoader.class);
		StageActivity.stageListener = Mockito.mock(StageListener.class);

		int virtualWidth = project.getXmlHeader().virtualScreenWidth;
		int virtualHeight = project.getXmlHeader().virtualScreenHeight;

		int virtualWidthHalf = virtualWidth / 2;
		int virtualHeightHalf = virtualHeight / 2;

		OrthographicCamera camera = new OrthographicCamera();
		StageActivity.stageListener.cameraPositioner = new CameraPositioner(camera, virtualHeightHalf, virtualWidthHalf);
	}

	@Test
	public void testSetCameraFocusPoint() {
		Float horizontalValue = 15f;
		Float verticalValue = 10f;
		ActionFactory factory = sprite.getActionFactory();
		Formula horizontal = new Formula(horizontalValue);
		Formula vertical = new Formula(verticalValue);
		ScriptSequenceAction sequenceAction = new ScriptSequenceAction(sprite.getScript(0));
		Action action = factory.createSetCameraFocusPointAction(sprite, sequenceAction, horizontal, vertical);
		action.act(1.0f);

		assertEquals(StageActivity.stageListener.cameraPositioner.getHorizontalFlex(), horizontalValue);
		assertEquals(StageActivity.stageListener.cameraPositioner.getVerticalFlex(), verticalValue);
	}
}

File: catroid/src/androidTest/java/org/catrobat/catroid/test/content/actions/SetCameraFocusPointActionTest.kt
/*
 * Catroid: An on-device visual programming system for Android devices
 * Copyright (C) 2010-2020 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * An additional term exception under section 7 of the GNU Affero
 * General Public License, version 3, is available at
 * http://developer.catrobat.org/license_additional_term
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.catroid.test.content.actions

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.scenes.scene2d.Action
import org.catrobat.catroid.content.Sprite
import org.catrobat.catroid.formulaeditor.Formula
import org.catrobat.catroid.formulaeditor.FormulaElement
import org.catrobat.catroid.stage.StageActivity
import org.catrobat.catroid.stage.StageListener
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class SetCameraFocusPointActionTest {
    companion object {
        private const val SCREEN_WIDTH = 1080.0f
        private const val SCREEN_HEIGHT = 1920.0f
    }

    private lateinit var spriteToFocus: Sprite
    private lateinit var stageListener: StageListener

    @Before
    fun setUp() {
        StageActivity.stageListener = StageListener()
        stageListener = StageActivity.stageListener
        stageListener.camera = OrthographicCamera()
        stageListener.camera.position.set(0.0f, 0.0f, 0.0f)
        stageListener.setVirtualWidthHalf(SCREEN_WIDTH / 2)
        stageListener.setVirtualHeightHalf(SCREEN_HEIGHT / 2)
        spriteToFocus = Sprite("spriteToFocus")
        spriteToFocus.look.setPositionInUserInterfaceDimensionUnit(0.0f, 0.0f)
    }

    @Test
    fun testCameraMovingWithSprite() {
        createAction(0.0f, 0.0f).act(1.0f)
        setPositionAndFocusCamera(10.0f, 10.0f)
        assertEquals(
            spriteToFocus.look.xInUserInterfaceDimensionUnit,
            stageListener.camera.position.x)
        assertEquals(
            spriteToFocus.look.yInUserInterfaceDimensionUnit,
            stageListener.camera.position.y)
    }

    @Test
    fun testNoChangeInPositionOfOtherSprite() {
        val sprite = Sprite("testSprite")
        sprite.look.setPositionInUserInterfaceDimensionUnit(200.0f, 100.0f)
        createAction(0.0f, 0.0f).act(1.0f)
        setPositionAndFocusCamera(10.0f, 10.0f)
        assertEquals(200.0f, sprite.look.xInUserInterfaceDimensionUnit)
        assertEquals(100.0f, sprite.look.yInUserInterfaceDimensionUnit)
    }

    @Test
    fun testNoChangeWithinHorizontalFlex() {
        createAction(50.0f, 0.0f).act(1.0f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.5f, 0.0f)
        assertEquals(Vector3(0.0f, 0.0f, 0.0f), StageActivity.stageListener.camera.position)
    }

    @Test
    fun testChangeOutsideHorizontalFlex() {
        createAction(50.0f, 0.0f).act(1.0f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.6f, 0.0f)
        assertEquals(
            Vector3(SCREEN_WIDTH / 2 * 0.1f, 0.0f, 0.0f),
            stageListener.camera.position)
    }

    @Test
    fun testNoChangeWithinVerticalFlex() {
        createAction(0.0f, 50.0f).act(1.0f)
        setPositionAndFocusCamera(0.0f, SCREEN_HEIGHT / 2 * 0.5f)
        assertEquals(
            Vector3(0.0f, 0.0f, 0.0f),
            stageListener.camera.position)
    }

    @Test
    fun testChangeOutsideVerticalFlex() {
        createAction(0.0f, 50.0f).act(1.0f)
        setPositionAndFocusCamera(0.0f, SCREEN_HEIGHT / 2 * 0.6f)
        assertEquals(
            Vector3(0.0f, SCREEN_HEIGHT / 2 * 0.1f, 0.0f),
            stageListener.camera.position)
    }

    @Test
    fun testChangeOutSideHorizontalAndVerticalFlex() {
        createAction(50.0f, 50.0f).act(1.0f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.6f, SCREEN_HEIGHT / 2 * 0.6f)
        assertEquals(
            Vector3(SCREEN_WIDTH / 2 * 0.1f, SCREEN_HEIGHT / 2 * 0.1f, 0.0f),
            stageListener.camera.position)
    }

    @Test
    fun testNoCameraChangeMultiplePositionChanges() {
        createAction(50.0f, 50.0f).act(1.0f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.4f, SCREEN_HEIGHT / 2 * 0.5f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.2f, SCREEN_HEIGHT / 2 * 0.1f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.3f, SCREEN_HEIGHT / 2 * 0.3f)
        setPositionAndFocusCamera(SCREEN_WIDTH / 2 * 0.5f, SCREEN_HEIGHT / 2 * 0.5f)
        assertEquals(
            Vector3(0.0f, 0.0f, 0.0f),
            stageListener.camera.position)
    }

    private fun createAction(horizontalFlex: Float, verticalFlex: Float): Action {
        return spriteToFocus.actionFactory.createSetCameraFocusPointAction(
            spriteToFocus,
            Formula(FormulaElement(FormulaElement.ElementType.NUMBER, horizontalFlex.toString(), null)),
            Formula(FormulaElement(FormulaElement.ElementType.NUMBER, verticalFlex.toString(), null))
        )
    }

    private fun setPositionAndFocusCamera(x: Float, y: Float) {
        spriteToFocus.look.setPositionInUserInterfaceDimensionUnit(x, y)
        stageListener.calculateCameraPositionForFocusedSprite()
    }
}
